### ISSUE TITLE
Proposal: an option `ecmaFeatures: {modulesDynamic: true}`

### ISSUE THREAD

**Issue by mysticatea** [eslint/espree#171]
Currently, espree needs `modules` option before parsing.
But it looks wonder to me. Espree seems to be possible to change modules flag [here](https://github.com/eslint/espree/blob/master/espree.js#L5147-5156).
This proposal adds an option `modulesDynamic`.
If this option turned on, when `import/export` were found, it switches `sourceType` to `modules`.
If the `es6` environment of ESLint includes this flag, I guess a problem that we cannot specify modules flag with CLI options is solved.

I tried to think reasons that espree needs `modules` option before parsing, but I didn't find.
What are the reasons?

**Comment by nzakas** [comment]
We need to know if the source is a module so we can enforce strict mode pre-checks during parsing. 

When native support for modules arrive in Node.js and the browser, you will also have to specify that the file is a module before parsing/executing.

**Comment by mysticatea** [comment]
> We need to know if the source is a module so we can enforce strict mode pre-checks during parsing.

`import/export` are appeared at same timing with the global `"use strict"` directive, so I guess we can handle as same as the global `"use strict"` directive.
...I thought so yesterday, but today I realized a source code might not have `import/export`.

foo.js:

```js
console.log("foo");
```

bar.js

```
import "./foo.js";
```

In this case, `foo.js` is in module, but a hint is nothing in `foo.js`. So we need a flag before parsing.

> When native support for modules arrive in Node.js and the browser, you will also have to specify that the file is a module before parsing/executing.

Oh, I got it. I had not known it... 

Thank you very much!
